File: 백준/Gold/2098. 외판원 순회/외판원 순회.cc
#include <bits/stdc++.h>
using namespace std;
const int INF = 987654321;
int n, dp[16][1 << 16], w[16][16];

int go(int here, int visited)
{
  if (visited == (1 << n) - 1)
  {
    return w[here][0] ? w[here][0] : INF;
  }
  int &ret = dp[here][visited];
  if (ret != -1)
    return ret;
  ret = INF;

  for (int i = 1; i < n; i++)
  {
    if (w[here][i] == 0 || visited & 1 << i)
      continue;
    ret = min(ret, go(i, visited | 1 << i) + w[here][i]);
  }
  return ret;
}
int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  cin >> n;
  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < n; j++)
    {
      cin >> w[i][j];
    }
  }

  memset(dp, -1, sizeof(dp));

  cout << go(0, 1) << "\n";

  return 0;
}

File: 백준/Gold/2098. 외판원 순회/외판원 순회.java

import java.util.*;

public class Main {
    static final int INF = 987654321;
    static int n;
    static int[][] a;
    static int[][] dp;

    static int go(int here, int visited) {
        if (visited == (1 << n) - 1) {
            return (a[here][0] != 0) ? a[here][0] : INF;
        }
        if (dp[here][visited] != -1)
            return dp[here][visited];
        int ret = INF;
        for (int next = 0; next < n; next++) {
            if (a[here][next] == 0)
                continue;
            if ((visited & (1 << next)) != 0)
                continue;
            ret = Math.min(ret, go(next, visited | (1 << next)) + a[here][next]);
        }
        return dp[here][visited] = ret;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        a = new int[n][n];
        dp = new int[n][(1 << n)];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = sc.nextInt();
            }
        }
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }
        System.out.println(go(0, 1));
        sc.close();
    }
}
